Plugin system for stt, tts and audioservices

forslund opened the pull request:
## Description
This adds a simple plugin system for loading external stt, tts and audioservices. The "plugins" are pip install-able modules presenting one or more entrypoints with a entrypoint group defined in setup.py ([Example](https://github.com/forslund/mycroft-tts-plugin-gtts/blob/master/setup.py))

TTS group: "mycroft.plugin.tts"
STT group: "mycroft.plugin.stt"
Audioservice group: "mycroft.plugin.audioservice"
Wake-word group: "mycroft.plugin.wake_word"

STT and TTS loads a single object for the specified module (as specified in the config), while audioservice loads all found audioservices and append to the list of builtins services. 

This system makes it easier to quickly add third-party services without changing core, it also makes it easier for users to select things like Polly where the requirement will automatically be installed together with the TTS module.

The detection system is based on the documentation for python packaging. [Link](https://packaging.python.org/guides/creating-and-discovering-plugins/)

-Future work: hotword engines. Can the same system be leveraged for skills? Update and reloading needs to be considered...-
@jarbasAl contributed an implementation of hot/wake-word engine plugins. There is a test plugin available [here](https://github.com/JarbasAl/dummy_wakeword_plugin)

## How to test
`mycroft-pip install git+https://github.com/forslund/mycroft-tts-plugin-gtts.git`

Then set the tts module to `google_tts_plug` assert that the google tts voice is heard

## Contributor license agreement signed?
CLA [ Yes ]

devops-mycroft commented:
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-2594))

devops-mycroft commented:
Voight Kampff Integration Test Failed ([Results](https://reports.mycroft.ai/core/PR-2594))

devops-mycroft commented:
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-2594))

krisgesling commented:
Hey sorry it's taken some time to get back to this. I really like it though. 

I think it will be great if we can keep mycroft-core to what is actually "core" to running any instance of Mycroft, and then provide the tooling to build up what's needed on top of that for specific implementations. 

The example plugin doesn't seem to be working for me. Looks like `gtts.lang.tts_langs` was added in [gTTS v2.0.0](https://github.com/pndurette/gTTS/pull/108) but the plugin is installing v1.1.7? I tried a quick requirement update in line with core (v2.0.4) but no luck and haven't dug into that further. 

@chrisveilleux if you agree with the addition of the plugin system, it would be good to merge it in pretty soon so it can be tested in the wild. Even though this isn't a breaking change itself, it will mean for 20.08 we can confidently remove non-core services and provide them as plugins instead.

forslund commented:
You're right that the requirement was a bit old. I've now updated it to track core's and for me it works [TM].

Do you get any error in the audio.log?

If the plug is loaded it should show
```
2020-07-29 07:56:47.651 | INFO     |  8201 | mycroft.tts.tts:create:524 | Loaded plugin google_tts_plug
```

krisgesling commented:
Hooray, working now. I think it was using a cached wheel when I tried to update it.

chrisveilleux commented on mycroft/audio/audioservice.py:
```diff
@@ -105,25 +131,54 @@ def load_services(config, bus, path=None):
     return service
 
 
+def load_plugins(config, bus):
```
could this be moved to utils.plugins?  `load_plugin()` is there, why not this?

forslund replied:
This loads plugins slightly differently from the TTS and STT since the audioservices are complete modules and can be loaded in a couple of different ways. This uses the mechanics defined in the util library and adapts it according to the needs of this specific service.

Adding a `load_audioservice_plugin()` to the general `mycroft.util.plugins` does to me seem a bit weird.

chrisveilleux commented on mycroft/stt/__init__.py:
```diff
@@ -547,6 +548,10 @@ def get_response(self, response):
         return response.json()["result"]["hypotheses"][0]["transcript"]
 
 
+def load_stt_plugin(module_name):
```
why a helper function here?  couldn't this one line of code be executed on line 581?

forslund replied:
Basically for keeping the `create()` method below clear and readable

chrisveilleux commented on mycroft/tts/tts.py:
```diff
@@ -464,6 +465,10 @@ def get_tts_class(self):
         pass
 
 
+def load_tts_plugin(module_name):
```
what is the purpose of this helper function? it is only executed once.

forslund replied:
For keeping the code below more readable, but I'll see if I can think of another way to keep it clear.

devops-mycroft commented:
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-2594))

chrisveilleux reviewed (dismissed):
I like the idea of a plugin system to keep the core code cleaner.  

@krisgesling Our documentation should be updated to explain how to use this.  I would prefer not to deploy it until the documentation is ready.

Some sort of automated testing should be added for this code.

forslund commented on mycroft/audio/audioservice.py:
```diff
@@ -82,16 +83,41 @@ def get_services(services_folder):
 
-        Returns:
-            List of started services.
+def load_builtins(config, bus, path=None):
```
internal?

forslund replied:
will change to something like `load_internal_services()`

forslund replied:
Change applied

forslund commented on mycroft/audio/audioservice.py:
```diff
@@ -105,25 +131,54 @@ def load_services(config, bus, path=None):
+    Returns:
+        List of started services.
+    """
+    builtins = load_builtins(config, bus, path)
+    plugins = load_plugins(config, bus)
+    return builtins + plugins
```
The loading is a rare occurrence and the objects are few, I think the clarity of doing the add instead of the extend is worth it here.

forslund replied:
I did refactor this a bit but I've kept the `+`  operation to generate a new list.

forslund commented on mycroft/audio/audioservice.py:
```diff
@@ -105,25 +131,54 @@ def load_services(config, bus, path=None):
+    """
+    services = []
+    plugins = find_plugins('mycroft.plugin.audioservice')
+    for p in plugins:
```
Are you sure that is a good rule to keep. Short variable names for short scopes are quite convenient and simple to follow.

forslund replied:
I have changed according to the suggestion.

forslund commented on mycroft/audio/audioservice.py:
```diff
@@ -82,16 +83,41 @@ def get_services(services_folder):
 
-        Returns:
-            List of started services.
+def load_builtins(config, bus, path=None):
```
Yes that can absolutely be done, same is true for STT and TTS

forslund commented on mycroft/audio/audioservice.py:
```diff
@@ -82,16 +83,41 @@ def get_services(services_folder):
+        try:
+            return service_module.autodetect(config, bus)
+        except Exception as e:
+            LOG.error('Failed to autodetect. ' + repr(e))
```
I was about to change this but when going through the code I realized these were intentionally set to error. This is since these are expected to fail to load if the prerequisite isn't installed so we don't want to clog up the log with traces for modules that aren't installed.

I'd like to wait to change this until third party optional modules are removed from core and all services are expected to load.

krisgesling replied:
I added a comment to the relevant issue for extracting optional services so we don't forget about this.

forslund commented:
I've updated the code styling changes requested and tried to answer the comments. I still think the small wrapper functions makes sense to have from a readability point of view but if you really like them gone I can inline them.

I'll write some unit tests and an entry for the documentation this weekend to go with this update.

devops-mycroft commented:
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-2594))

devops-mycroft commented:
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-2594))

krisgesling commented:
Documentation for the plugin system is now available at: 
https://mycroft-ai.gitbook.io/docs/mycroft-technologies/mycroft-core/plugins

devops-mycroft commented:
Voight Kampff Integration Test Failed ([Results](https://reports.mycroft.ai/core/PR-2594))

forslund commented:
I've rebased this and fixed the conflict. I've also added some unittests for the plugins util. Documentation is available here: https://mycroft-ai.gitbook.io/docs/mycroft-technologies/mycroft-core/plugins

devops-mycroft commented:
Voight Kampff Integration Test Failed ([Results](https://reports.mycroft.ai/core/PR-2594))

forslund commented:
Hmm, must have messed up the rebase in audioservice...will invesigate

devops-mycroft commented:
Voight Kampff Integration Test Failed ([Results](https://reports.mycroft.ai/core/PR-2594))

devops-mycroft commented:
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-2594))

JarbasAl commented:
I'm thinking we need this for wake words and enclosures

forslund commented:
Wakewords is something I've planned for as well but wanted to make as a follow up PR since this is enough to review.

Enclosures is a nice idea I didn't think of. Previously the idea was to extract the enclosure system from core entirely and just use the messagebus connection for it (mycroft-messagebus-client + a base enclosure module) but including it as a pluggable system should work quite well also.

devops-mycroft commented:
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-2594))

krisgesling reviewed (approved):
This all looks great, working as expected, and I've updated the docs for the addition of Wake Word Plugins:
https://mycroft-ai.gitbook.io/docs/mycroft-technologies/mycroft-core/plugins
